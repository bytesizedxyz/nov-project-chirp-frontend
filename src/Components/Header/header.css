/* App styling */
.App {
  text-align: center;
  height: 100vh;
  width: 100vw;
  color: #fff;
}

/* theme button */
.theme-button {
  border: 0px;
}

.theme-button :focus {
  outline-style: none;
  box-shadow: none;
  border-color: transparent;
}

/* Light Theme */
.lightBlueBackground {
  background-color: #dff6ff;
}

.eggshellBackground {
  background-color: #f0ead6;
}

.greyBackground {
  background-color: #9c9999;
}

.blueBackground {
  background-color: #2f67f8;
  /* background-color: #2049d9; */
}

.brownFont {
  color: #581f11;
}

.eggshellFont {
  color: #f0ead6;
}

.eggshellBorder {
  border: 2px solid #f0ead6;
}

.searchIconBorderEggshell {
  border-left: solid 2px #f0ead6;
  border-bottom: solid 2px #f0ead6;
  border-top: solid 2px #f0ead6;
}

/* Dark Theme */
.brownBackground {
  background-color: #292323;
}

.greyBackground {
  background-color: #9c9999;
}

.blackBackground {
  background-color: black;
}

.whiteFont {
  color: #fff;
}

.redFont {
  color: #990303;
}

.blackFont {
  color: black;
}

.redBorder {
  border: 2px solid #990303;
}

.searchIconBorderRed {
  border-left: solid 2px #990303;
  border-bottom: solid 2px #990303;
  border-top: solid 2px #990303;
}

/* placeholder font */
::placeholder {
  font-size: 14px;
  font-weight: 700;
  color: #f0ead6;
  opacity: 1;
}

:-ms-input-placeholder {
  font-size: 14px;
  font-weight: 700;
  color: #f0ead6;
  opacity: 1;
}

::-ms-input-placeholder {
  font-size: 14px;
  font-weight: 700;
  color: #f0ead6;
  opacity: 1;
}

.zIndex {
  z-index: 1;
}

.logo {
  height: 120px;
  width: 120px;
  margin-left: 10px;
  margin-top: 10px;
}

.chirpButton {
  font-weight: 700;
  font-size: 20px;
  height: 40px;
  width: 120px;
  border-radius: 2px;
}

.search {
  padding: 3px;
  border-radius: 0px 4px 4px 0px;
  height: 30px;
  width: 280px;
  font-size: 18px;
}

.profileImage {
  border-radius: 50%;
  margin-right: 10px;
  margin-top: 10px;
  opacity: 0.75;
}

.modal {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.modal-main {
  position: fixed;
  width: 80%;
  height: auto;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}

.modal-main textarea {
  margin-bottom: 3%;
  height: 10em;
}

.modal-main button {
  margin-right: 3%;
  margin-bottom: 3%;
}

.modalOpen {
  z-index: 1;
}

.display-none {
  display: none;
}

.textAreaFont {
  font-size: 18px;
}

.searchIcon {
  padding: 6px;
  font-size: 24px;
  font-weight: 500;
  border-radius: 4px 0px 0px 4px;
}

h1 {
  font-size: 80px;
  margin: 0px;
  font-family: "Bowlby One SC", cursive;
  letter-spacing: 2px;
}

.headerBottom {
  margin-top: -15px;
  margin-bottom: 15px;
  width: 50%;
}

/* font */
@import url("https://fonts.googleapis.com/css?family=Bowlby+One+SC");

/* flexbox */
.flexRowAround {
  display: flex;
  justify-content: space-around;
  align-items: center;
}

.flexColumnAround {
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  align-items: center;
  width: 100%;
}

.flexRowBetween {
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: 100%;
}

.flexRowCenter {
  display: flex;
  justify-content: center;
  align-items: center;
}

/* media queries */
@media only screen and (min-width: 300px) {
  .modal-main textarea {
    width: 60%;
  }
}

@media only screen and (min-width: 600px) {
  .modal-main textarea {
    height: 10em;
  }
}

@media only screen and (min-width: 1000px) {
  .modal-main textarea {
    height: 10em;
  }
}
