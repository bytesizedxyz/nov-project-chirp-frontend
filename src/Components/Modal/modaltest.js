// import React from "react";
// import { cleanup, render, fireEvent, wait } from "react-testing-library";

// // this adds custom jest matchers from jest-dom
// import "jest-dom/extend-expect";

// //needed components to render properly
// import Modal from "./Modal";
// import { ThemeContext, themes } from "../../ThemeProvider";
// import { library } from "@fortawesome/fontawesome-svg-core";
// import { faPlus, faSearch } from "@fortawesome/free-solid-svg-icons";
// const { toMatchDiffSnapshot } = require("snapshot-diff");
// library.add(faPlus, faSearch);

// afterEach(cleanup);

// describe("Modal Rendering", () => {
//   it("renders the modal with props passed to it", async () => {
//     const open = false;
//     const hideModal = jest.fn();
//     const closeAndSend = jest.fn();

//     const { getByText } = render(
//       <Modal open={open} handleClose={hideModal} addPost={closeAndSend} />
//     );

//     const modalText = getByText("Add New Post");

//     expect(modalText).toHaveTextContent("Add New Post");
//   });
// });
