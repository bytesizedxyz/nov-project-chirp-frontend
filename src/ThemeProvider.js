import React from "react";

export const themes = {
  light: {
    lightBlueBackground: "lightBlueBackground",
    eggshellBackground: "eggshellBackground",
    greyBackground: "greyBackground",
    blueBackground: "blueBackground",
    whiteFont: "whiteFont",
    eggshellFont: "eggshellFont",
    brownFont: "brownFont",
    eggshellBorder: "eggshellBorder",
    searchIconBorder: "searchIconBorderEggshell",
    obiTwitter: "Obi Wan Kenobi Twitter",
    iconColor:"#fff"
  },
  dark: {
    brownBackground: "brownBackground",
    greyBackground: "greyBackground",
    whiteFont: "whiteFont",
    blackFont: "blackFont",
    redFont: "redFont",
    redBorder: "redBorder",
    searchIconBorder: "searchIconBorderRed",
    blackBackground: "blackBackground",
    darthTwitter: "Darth Twitter",
    iconColor: "#990303"
  }
};

export const ThemeContext = React.createContext({
  theme: themes.dark,
  toggleTheme: () => {} // default value
});
