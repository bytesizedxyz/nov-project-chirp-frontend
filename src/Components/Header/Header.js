import React, { Component } from 'react';
import {
  Button,
  Modal,
  TextArea,
  Input,
  Menu,
  Header,
  Grid,
  Icon
} from 'semantic-ui-react';
import Gravatar from 'gravatar-react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import PropTypes from 'prop-types';
import { ThemeContext, themes } from '../../ThemeProvider';
import EmpireLogo from '../empire-brands';
import './header.css';

class NavBar extends Component {
  constructor() {
    super();

    this.state = {
      message: '',
      modalOpen: false
    };

    this.handleChange = this.handleChange.bind(this);
    this.closeAndSend = this.closeAndSend.bind(this);
  }

  handleOpen = () => this.setState({ modalOpen: true });

  handleClose = () => this.setState({ modalOpen: false });

  handleChange(e) {
    const { name, value } = e.target;
    this.setState({
      [name]: value
    });
  }

  closeAndSend() {
    const { addPost } = this.props;
    const { message } = this.state;
    console.log('running close and send');
    this.setState({ modalOpen: false, message: '' });
    addPost(message);
  }

  render() {
    const { user, handleFilter, filter } = this.props;
    const { message } = this.state;
    const email = user || 'email@gmail.com';
    const GravatarBlock = (
      <Gravatar
        email={email}
        size={240}
        rating="PG"
        alt="Profile Avatar"
        default="monsterid"
        secure
      />
    );
    return (
      <ThemeContext.Consumer>
        {({ theme, toggleTheme }) => (
          <Menu
            color={theme.currentTheme === 'dark' ? 'black' : 'blue'}
            inverted
            borderless={false}
            fluid
            size="huge"
            widths={3}
            className={`header`}
          >
            <Menu.Item onClick={toggleTheme}>
              <Icon
                color={theme.currentTheme === 'dark' ? 'red' : 'white'}
                size="massive"
                name={theme.currentTheme === 'dark' ? 'empire' : 'rebel'}
              />
            </Menu.Item>
            <Menu.Item>
              <Grid.Row>
                <Header
                  size="huge"
                  as="h1"
                  color={theme.currentTheme === 'dark' ? 'red' : 'white'}
                >
                  {theme.currentTheme === 'dark'
                    ? 'Darth Twitter'
                    : 'Obi Wan Twitter'}
                </Header>
                <Modal
                  centered
                  trigger={
                    <Button
                      size="medium"
                      onClick={this.handleOpen}
                      data-testid="addPostButton"
                      style={{ marginRight: '20px' }}
                      color={theme.currentTheme === 'dark' ? 'red' : 'blue'}
                    >
                      <FontAwesomeIcon icon="plus" />
                    </Button>
                  }
                  open={this.state.modalOpen}
                  onClose={this.handleClose}
                  closeIcon
                  data-testid="modal"
                >
                  <Modal.Header
                    style={
                      theme.currentTheme === 'dark'
                        ? { backgroundColor: 'black' }
                        : { backgroundColor: 'blue' }
                    }
                  >
                    <h1
                      style={
                        theme.currentTheme === 'dark'
                          ? { color: 'red', textAlign: 'center' }
                          : { color: 'white', textAlign: 'center' }
                      }
                    >
                      Add New Post
                    </h1>
                  </Modal.Header>
                  <Modal.Content
                    style={
                      theme.currentTheme === 'dark'
                        ? { backgroundColor: 'black' }
                        : { backgroundColor: 'blue' }
                    }
                  >
                    <TextArea
                      placeholder="Tell us more"
                      data-testid="addPostText"
                      onChange={this.handleChange}
                      name="message"
                      autoHeight
                      rows={10}
                      style={{ width: '100%' }}
                      value={message}
                    />
                  </Modal.Content>
                  <Modal.Actions
                    style={
                      theme.currentTheme === 'dark'
                        ? { backgroundColor: 'black', textAlign: 'center' }
                        : { backgroundColor: 'blue', textAlign: 'center' }
                    }
                  >
                    <Button onClick={this.handleClose}>Close</Button>
                    <Button onClick={this.closeAndSend}>Submit</Button>
                  </Modal.Actions>
                </Modal>
                <Input
                  onChange={handleFilter}
                  placeholder="Search for chirps"
                  type="text"
                  name={filter}
                  action="Search"
                  size="medium"
                />
              </Grid.Row>
            </Menu.Item>
            <Menu.Item className="profileImage">{GravatarBlock}</Menu.Item>
          </Menu>
        )}
      </ThemeContext.Consumer>
    );
  }
}

NavBar.propTypes = {
  addPost: PropTypes.func.isRequired,
  handleFilter: PropTypes.func.isRequired,
  user: PropTypes.objectOf(PropTypes.string).isRequired,
  filter: PropTypes.string.isRequired
};

export default NavBar;
