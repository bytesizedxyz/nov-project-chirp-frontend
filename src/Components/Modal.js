import React from 'react';
import { ThemeContext } from '../ThemeProvider';

const Modal = ({ addPost, handleClose, open, children }) => {
  const showHideClassName = open ? 'modalOpen' : 'modal display-none';
  return (
    <ThemeContext.Consumer>
      {({ theme }) => (
        <div className={showHideClassName}>
          <section
            className={`modal-main
            ${theme.whiteFont}
            ${theme.brownBackground}
            ${theme.blueBackground}
            `}
          >
            <div>{children}</div>
            <button
              className={`chirpButton
              ${theme.whiteFont}
              ${theme.brownBackground}
              ${theme.blueBackground}
              ${theme.eggshellBorder}
              ${theme.eggshellFont}
              ${theme.redFont}
              ${theme.redBorder}
              `}
              onClick={handleClose}
            >
              close
            </button>
            <button
              className={`chirpButton
              ${theme.whiteFont}
              ${theme.brownBackground}
              ${theme.blueBackground}
              ${theme.eggshellBorder}
              ${theme.redBorder}
              ${theme.eggshellFont}
              ${theme.redFont}
              `}
              onClick={addPost}
            >
              submit
            </button>
          </section>
        </div>
      )}
    </ThemeContext.Consumer>
  );
};

export default Modal;
