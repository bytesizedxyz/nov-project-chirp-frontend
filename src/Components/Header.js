import React, { Component } from "react";
import Gravatar from "gravatar-react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import EmpireLogo from "./empire-brands";
import Modal from "./Modal";
import { ThemeContext } from "../ThemeProvider";
import "./header.css";

class Header extends Component {
  state = {
    message: "",
    open: false
  };

  showModal = () => {
    this.setState({ show: true });
  };

  hideModal = () => {
    this.setState({ show: false });
  };

  handleChange = e => {
    const { name, value } = e.target;
    this.setState({
      [name]: value
    });
  };

  closeAndSend = () => {
    console.log("running close and send");
    this.setState({ open: false, message: "" });
    this.props.addPost(this.state.message);
  };

  handleSearch = () => {
    console.log("Searching for new chirps");
  };

  render() {
    const { user, handleFilter, filter } = this.props;
    let GravatarBlock;
    if (user && user.email) {
      GravatarBlock = (
        <ThemeContext.Consumer>
          {({ theme }) => (
            <Gravatar
              className={`profileImage
              ${theme.redBorder}
              ${theme.eggshellBorder}`}
              email={this.props.user.email}
              size={120}
              rating="PG"
              alt="Profile Avatar"
              default="monsterid"
              secure
            />
          )}
        </ThemeContext.Consumer>
      );
    }

    return (
      <ThemeContext.Consumer>
        {({ theme, toggleTheme }) => (
          <div
            className={`flexColumnAround
            ${theme.brownBackground}
            ${theme.blueBackground}`}
          >
            <div
              className={`flexRowBetween
              ${theme.brownBackground}
              ${theme.blueBackground}`}
            >
              <div
                className={`theme-button ${theme.brownBackground} ${theme.blueBackground}`}
                onClick={toggleTheme}
              >
                <EmpireLogo />
              </div>
              <h1 className={`${theme.eggshellFont} ${theme.blackFont}`}>
                {theme.darthTwitter}
                {theme.obiTwitter}
              </h1>
              {GravatarBlock}
            </div>

            <div
              className={`headerBottom flexRowAround
              ${theme.brownBackground}
              ${theme.blueBackground}`}
            >
              <Modal
                open={this.state.show}
                handleClose={this.hideModal}
                addPost={this.closeAndSend}
              >
                <h1 className={`${theme.blackFont} ${theme.eggshellFont}`}>Add New Post</h1>
                <textarea
                  value={this.state.message}
                  onChange={this.handleChange}
                  rows="5"
                  cols="50"
                  type="text"
                  name="message"
                  maxLength="280"
                  className={`greyBackground whiteFont textAreaFont`}
                />
              </Modal>
              <button
                className={`chirpButton
                ${theme.brownBackground}
                ${theme.redBorder}
                ${theme.redFont}
                ${theme.eggshellFont}
                ${theme.font}
                ${theme.blueBackground}
                ${theme.eggshellBorder}`}
                onClick={this.showModal}
              >
                <FontAwesomeIcon icon="plus" />
              </button>
              <span className="flexRowCenter">
                <FontAwesomeIcon
                  icon="search"
                  className={`searchIcon ${theme.searchIconBorder} ${theme.redFont}`}
                />
                <input
                  onChange={handleFilter}
                  placeholder="Search for chirps"
                  className={`search
                  ${theme.brownBackground}
                  ${theme.blueBackground}
                  ${theme.eggshellFont}
                  ${theme.eggshellBorder}
                  ${theme.whiteFont}
                  ${theme.redBorder}`}
                  type="text"
                  name={filter}
                />
              </span>
              <button
                className={`chirpButton
                  ${theme.brownBackground}
                  ${theme.blueBackground}
                  ${theme.eggshellFont}
                  ${theme.eggshellBorder}
                  ${theme.whiteFont}
                  ${theme.redFont}
                  ${theme.redBorder}`}
                onClick={this.handleSearch}
              >
                SEARCH
              </button>
            </div>
          </div>
        )}
      </ThemeContext.Consumer>
    );
  }
}

export default Header;
