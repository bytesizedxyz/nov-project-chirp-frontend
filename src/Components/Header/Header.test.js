import React from "react";
import { cleanup, render, fireEvent, wait } from "react-testing-library";

// this adds custom jest matchers from jest-dom
import "jest-dom/extend-expect";

//needed components to render properly
import Header from "./Header";
import Modal from "../Modal";
import { ThemeContext, themes } from "../../ThemeProvider";
import { library } from "@fortawesome/fontawesome-svg-core";
import { faPlus, faSearch } from "@fortawesome/free-solid-svg-icons";
const { toMatchDiffSnapshot } = require("snapshot-diff");
expect.extend({ toMatchDiffSnapshot });
library.add(faPlus, faSearch);

afterEach(cleanup);

xdescribe("Header functionality", () => {
  xdescribe("showModal", () => {
    it("should setState", () => {
      //making the header instance
      const component = new Header();
      //redefining the setstate function
      component.setState = jest.fn();

      //opening the modal
      component.showModal();
      expect(component.setState).toHaveBeenCalledWith({ show: true });

      //closing the modal
      component.hideModal();
      expect(component.setState).toHaveBeenCalledWith({ show: false });
    });
  });

  it("it renders the header components such as title, search input, add button and grab user image", async () => {
    //defining simulated props
    const user = {
      attributes: {
        email: "email@gmail.com"
      }
    };
    const filter = "";

    //rendering header
    const { getByText, getByPlaceholderText, getByAltText, getByTestId } = render(
      <Header user={user.attributes} filter={filter} />
    );

    //Rendered items to look for
    const header = getByText("Darth Twitter");
    const search = getByPlaceholderText("Search for chirps");
    const addButton = getByTestId("addPostButton");
    const image = getByAltText("Profile Avatar");

    //event that are expected to be true
    expect(header.innerHTML).toBe("Darth Twitter");
    expect(search).toBeInTheDocument();
    expect(search.value).toBe("");
    expect(addButton).toBeInTheDocument();
    expect(image).toBeVisible();

    //change value of input
    fireEvent.change(search, { target: { value: "cross" } });
    //evaluate that the value has been changed for the input
    expect(search.value).toBe("cross");
  });

  /*
  TODO: 
    [ ] - Make header change theme, might have to move it up into the app test instead
  */
  it("can open add post modal and able to click and change theme", async () => {
    //defining simulated props
    const toggleTheme = jest.fn();
    const theme = themes.dark;
    const themeChange = {
      theme,
      toggleTheme: toggleTheme
    };

    //rending header with themecontext provider
    const { getByTestId, asFragment } = render(
      <ThemeContext.Provider value={themeChange}>
        <Header />
      </ThemeContext.Provider>
    );

    //Rendered items to look for
    const addButton = getByTestId("addPostButton");
    const empireLogo = getByTestId("SVGIcon");

    //generating snapshot to test if modal has been opened
    const firstRender = asFragment();
    //clicking add modal to see if anything has changed
    fireEvent.click(addButton);
    //comparing the two snapshots
    expect(firstRender).toMatchDiffSnapshot(asFragment());

    //event that are expected to be true
    expect(empireLogo).toBeInTheDocument();
    expect(empireLogo).toHaveClass("brownBackground");

    // clicking the logo to change the theme
    fireEvent.click(empireLogo);
    expect(toggleTheme).toHaveBeenCalledTimes(1);

    // await wait(() => expect(empireLogo).toHaveClass("blueBackground"));
  });
  it("can change the value of the modal textarea and submit it", async () => {
    const addPost = jest.fn();
    const closeAndSend = jest.fn();
    const { getByTestId, getByText, asFragment } = render(
      <Header addPost={addPost}>
        <Modal addPost={closeAndSend} />
      </Header>
    );
    //generating first snapshot of render
    // const firstRender = asFragment();

    //rendered items to look for
    const modal = getByTestId("modal");
    const submit = getByText("submit");
    const textarea = getByTestId("addPostText");

    //events expected to be true
    expect(modal).toContainHTML("Add New Post");
    expect(modal).toHaveTextContent("submit");

    //testing default value of text area
    expect(textarea.value).toBe("");
    //change value of input
    fireEvent.change(textarea, { target: { value: "This is going to be a new post" } });
    //evaluate that the value has been changed for the input
    expect(textarea.value).toBe("This is going to be a new post");
    // expect(firstRender).toMatchDiffSnapshot(asFragment());
    
    //clicking submit button
    fireEvent.click(submit);
    expect(addPost).toHaveBeenCalledTimes(1);
  });
});
