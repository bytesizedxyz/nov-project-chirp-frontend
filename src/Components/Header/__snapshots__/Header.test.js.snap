// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar functionality can open add post modal and able to click and change theme 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;
